pkg/controller/agent/pod.go

```diff
@@ -50,63 +63,197 @@ var (
+			ConfigSecretName(params.Agent.Name),
+			FleetSetupVolumeName,
+			FleetSetupMountPath,
+			FleetSetupFileName,
+			0440,
 		)
-		vols = append(vols, caVolume)
+		vols = append(vols, fleetSetupConfigVol)
+
+		if spec.FleetServerEnabled {
+			// ECK creates CA and a certificate for Fleet Server to use. This volume contains those.
+			fleetCAVolume := volume.NewSecretVolumeWithMountPath(
+				fleetCerts.Name,
+				FleetCertsVolumeName,
+				FleetCertsMountPath,
+			)
+			vols = append(vols, fleetCAVolume)
+
+			// Beats managed by the Elastic Agent does not trust the Elasticsearch CA that Elastic Agent itself
+			// is configured to trust. There is currently no way to configure those Beats to trust a particular CA.
+			// The intended way to handle this is to allow Fleet to provide Beat output settings, but due to
+			// https://github.com/elastic/kibana/issues/102794 this is not supported outside of UI. To workaround this
+			// limitation the Agent is going to update Pod-wide CA store before starting Elastic Agent.
+			esAssoc := association.GetAssociationOfType(params.Agent.GetAssociations(), commonv1.ElasticsearchAssociationType)
+			if esAssoc != nil {
+				builder = builder.
+					WithCommand([]string{"/usr/bin/env", "bash", "-c", trustCAScript(path.Join(certificatesDir(esAssoc), CAFileName))})
+			}
+		} else {
+			// See the long comment above.
+			// We are processing Elastic Agent resource. As the reference chain is:
+			// Elastic Agent ---> Fleet Server ---> Elasticsearch
+			// we need first to identify the Fleet Server.
+			fs, err := getFsAssociation(params)
+			if err != nil {
+				return corev1.PodTemplateSpec{}, err
+			}
+
+			// now we can find the Elasticsearch association and mount its CA Secret
+			esfsAssoc := association.GetAssociationOfType(params.Agent.GetAssociations(), commonv1.ElasticsearchAssociationType)
+			if esfsAssoc != nil {
+				caSecretName := esfsAssoc.AssociationConf().GetCASecretName()
+				caVolume := volume.NewSecretVolumeWithMountPath(
+					caSecretName,
+					fmt.Sprintf("%s-certs", esfsAssoc.AssociationType()),
+					certificatesDir(esfsAssoc),
+				)
+				vols = append(vols, caVolume)
+
+				// because of the reference chain (Elastic Agent ---> Fleet Server ---> Elasticsearch), we are going to get
+				// notified when CA of Elasticsearch changes as Fleet Server resource will get updated as well. But what we
+				// also need to do is to roll Elastic Agent Pods to pick up the update CA. To do be able to do that, we are
+				// adding Fleet Server associations (which includes Elasticsearch) to config hash attached to Elastic Agent
+				// Pods.
+				if err := commonassociation.WriteAssocsToConfigHash(params.Client, fs.GetAssociations(), configHash); err != nil {
+					return corev1.PodTemplateSpec{}, err
+				}
+
+				builder = builder.
+					WithCommand([]string{"/usr/bin/env", "bash", "-c", trustCAScript(path.Join(certificatesDir(esfsAssoc), CAFileName))})
+			}
+		}
```

Comment:
Same here I think it would help readability to break it down into smaller helper functions otherwise it is quite hard to follow the large `if/else` clauses.